onadata/apps/api/viewsets/open_data_viewset.py

```diff
@@ -0,0 +1,185 @@
+                'alias': header
+            })
+
+        # using nested loops to determine what valid data types to set for
+        # tableau.
+        for header in self.xform_headers:
+            for quest_name, quest_type in self.flattened_dict.items():
+                if header == quest_name or header.endswith('_%s' % quest_name):
+                    append_to_tableau_colulmn_headers(header, quest_type)
+                    break
+            else:
+                append_to_tableau_colulmn_headers(header)
+
+        return tableau_colulmn_headers
+
+    @detail_route(methods=['GET'])
+    def data(self, request, **kwargs):
+        self.object = self.get_object()
+        # get greater than value and cast it to an int
+        gt = request.query_params.get('gt_id')
+        gt = gt and parse_int(gt)
+
+        data = []
+        if isinstance(self.object.content_object, XForm):
+            if not self.object.active:
+                return Response(status=status.HTTP_404_NOT_FOUND)
+
+            xform = self.object.content_object
+            qs_kwargs = {'xform': xform}
+            if gt:
+                qs_kwargs.update({'id__gt': gt})
+
+            instances = Instance.objects.filter(
+                **qs_kwargs
+            )[:self.MAX_INSTANCES_PER_REQUEST]
+            csv_df_builder = CSVDataFrameBuilder(
+                xform.user.username, xform.id_string, include_images=False
+            )
+            data = csv_df_builder._format_for_dataframe(
+                DataInstanceSerializer(instances, many=True).data,
+                key_replacement_obj={
+                    'pattern': r"(/|-|\[|\])",
+                    "replacer": r"_"
+                }
+            )
+
+        return Response(data)
+
+    def destroy(self, request, *args, **kwargs):
+        self.get_object().delete()
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+    @detail_route(methods=['GET'])
+    def schema(self, request, **kwargs):
+        self.object = self.get_object()
+        if isinstance(self.object.content_object, XForm):
+            xform = self.object.content_object
+            self.xform_headers = replace_special_characters_with_underscores(
+                xform.get_headers()
+            )
+
+            xform_json = json.loads(xform.json)
+            self.flatten_xform_columns(
+                json_of_columns_fields=xform_json.get('children')
+            )
+
+            tableau_column_headers = self.get_tableau_column_headers()
+
+            data = {
+                'column_headers': tableau_column_headers,
+                'connection_name': "%s_%s" % (
+                    xform.project_id, xform.id_string
+                ),
+                'table_alias': xform.title
+            }
+
+            return Response(data=data, status=status.HTTP_200_OK)
+
+        return Response(status=status.HTTP_404_NOT_FOUND)
+
+    @list_route(methods=['GET'])
+    def uuid(self, request, *args, **kwargs):
```

Comment:
This seems to indicate that any authenticated user can get access to any form data as long as they have an object_id. This should not be the case, this should be only available to an authenticated user with permissions to the form and data.